### PR TITLE
portable: allow caller to override extension-release name check

### PR DESCRIPTION
**Pull request by bluca**
When the --force flag is used, do not insist that the extension-release file has to match the extension image name

As requested in: https://github.com/systemd/systemd/pull/20281#issuecomment-948393538

### CODE REVIEW

File: src/basic/os-util.c [hunk 1]
```diff
@@ -36,7 +36,7 @@ bool image_name_is_valid(const char *s) {
         return true;
 }
 
-int path_is_extension_tree(const char *path, const char *extension) {
+int path_is_extension_tree(const char *path, const char *extension, bool extension_release_nonstrict) {
```
Comment (poettering) [inline src/basic/os-util.c]: I'd always prefer "positive" ways to name thins instead of "negated" ones. i.e. maybe name this "relax_extension_release_check" or so?

File: src/basic/os-util.c [hunk 2]
```diff
@@ -123,33 +123,35 @@ int open_extension_release(const char *root, const char *extension, char **ret_p
...
+                                        _cleanup_free_ char *extension_release_xattr = NULL;
+                                        k = fgetxattr_malloc(extension_release_fd, "user.extension-release.strict", &extension_release_xattr);
+                                        if (k < 0 && !ERRNO_IS_NOT_SUPPORTED(k) && k != -ENODATA)
+                                                log_debug_errno(k,
+                                                                "%s/%s: Failed to read 'user.extension-release.strict' extended attribute from file: %m",
+                                                                extension_release_dir_path, de->d_name);
+                                        if (k < 0) {
+                                                log_debug("%s/%s does not have user.extension-release.strict xattr, ignoring.", extension_release_dir_path, de->d_name);
+                                                continue;
+                                        }
```
Comment (poettering) [inline src/basic/os-util.c]: something is off here, as this will log twice about the same error in some cases.

File: src/basic/os-util.c [hunk 3]
```diff
@@ -123,33 +123,35 @@ int open_extension_release(const char *root, const char *extension, char **ret_p
...
+                                                log_debug_errno(k,
+                                                                "%s/%s: Failed to parse 'user.extension-release.strict' extended attribute from file: %m",
+                                                                extension_release_dir_path, de->d_name);
+                                        else if (k > 0)
+                                                log_debug("%s/%s: 'user.extension-release.strict' attribute is true, ignoring file.",
                                                         extension_release_dir_path, de->d_name);
-                                if (k < 0) {
-                                        log_debug("%s/%s does not have user.extension-release.strict xattr, ignoring.", extension_release_dir_path, de->d_name);
-                                        continue;
-                                }
```
Comment (poettering) [inline src/basic/os-util.c:136]: honestly, this check is now complex enough, maybe split the check for the xattr out into a helper function?

File: src/basic/os-util.c (hunk 2 above)
Comment (bluca) [reply src/basic/os-util.c]: pre-existing condition, but will fix

File: man/portablectl.xml [hunk 4]
```diff
@@ -378,7 +378,8 @@
         <term><option>--force</option></term>
 
         <listitem><para>Skip safety checks and attach or detach images (with extensions) without first ensuring
-        that the units are not running.</para></listitem>
+        that the units are not running, and do not insist that the extension-release.NAME file in the extension image
```
Comment (poettering) [inline man/portablectl.xml]: `<filename>` and `<replaceable>` tags could probably see some usage here...

File: src/basic/os-util.c [hunk 5]
```diff
@@ -58,7 +58,47 @@ int path_is_extension_tree(const char *path, const char *extension) {
...
+        assert(extension_release_dir_path);
+        assert(filename);
+
+        /* No xattr or cannot parse it? Then skip this. */
+        _cleanup_free_ char *extension_release_xattr = NULL;
+        r = fgetxattr_malloc(extension_release_fd, "user.extension-release.strict", &extension_release_xattr);
+        if (r < 0) {
+                if (!ERRNO_IS_NOT_SUPPORTED(r) && r != -ENODATA)
+                        log_debug_errno(r,
+                                        "%s/%s: Failed to read 'user.extension-release.strict' extended attribute from file: %m",
```
Comment (poettering) [inline src/basic/os-util.c]: also add `, ignoring: %m` to the end?

File: src/basic/os-util.c [hunk 6]
```diff
@@ -58,7 +58,47 @@ int path_is_extension_tree(const char *path, const char *extension) {
...
+
+        assert(extension_release_fd >= 0);
+        assert(extension_release_dir_path);
+        assert(filename);
+
+        /* No xattr or cannot parse it? Then skip this. */
+        _cleanup_free_ char *extension_release_xattr = NULL;
+        r = fgetxattr_malloc(extension_release_fd, "user.extension-release.strict", &extension_release_xattr);
+        if (r < 0) {
+                if (!ERRNO_IS_NOT_SUPPORTED(r) && r != -ENODATA)
```
Comment (poettering) [inline src/basic/os-util.c]: use ERRNO_XATTR_IS_ABSENT() for this (see other PR #24947)

File: src/basic/os-util.c [hunk 7]
```diff
@@ -58,7 +58,47 @@ int path_is_extension_tree(const char *path, const char *extension) {
...
+                        log_debug("%s/%s does not have user.extension-release.strict xattr, ignoring.", extension_release_dir_path, filename);
+
+                return r;
+        }
+
+        /* Explicitly set to request strict matching? Skip it. */
+        r = parse_boolean(extension_release_xattr);
+        if (r < 0)
+                log_debug_errno(r,
+                                "%s/%s: Failed to parse 'user.extension-release.strict' extended attribute from file: %m",
```
Comment (poettering) [inline src/basic/os-util.c]: add `, ignoring:`

File: src/core/namespace.c [hunk 8]
```diff
@@ -1382,7 +1382,7 @@ static int apply_one_mount(
                 if (isempty(host_os_release_id))
                         return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "'ID' field not found or empty in 'os-release' data of OS tree '%s': %m", empty_to_root(root_directory));
 
-                r = load_extension_release_pairs(mount_entry_source(m), extension_name, &extension_release);
+                r = load_extension_release_pairs(mount_entry_source(m), extension_name, false, &extension_release);
                 if (r == -ENOENT && m->ignore)
```
Comment (poettering) [inline src/core/namespace.c:1386]: `/* relax_extension_release_check= */`

File: src/shared/dissect-image.h [hunk 9]
```diff
@@ -208,6 +208,7 @@ typedef enum DissectImageFlags {
         DISSECT_IMAGE_MOUNT_IDMAPPED           = 1 << 19, /* Mount mounts with kernel 5.12-style userns ID mapping, if file system type doesn't support uid=/gid= */
         DISSECT_IMAGE_MANAGE_PARTITION_DEVICES = 1 << 20, /* Manage partition devices, e.g. probe each partition in more detail */
         DISSECT_IMAGE_BLOCK_DEVICE             = DISSECT_IMAGE_MANAGE_PARTITION_DEVICES,
+        DISSECT_IMAGE_OS_EXT_NON_STRICT        = 1 << 21, /* Don't insist that the extension-release file name matches the image name */
```
Comment (poettering) [inline src/shared/dissect-image.h]: DISSECT_IMAGE_RELAX_SYSTEM_EXTENSION_CHECK

File: src/shared/dissect-image.c [hunk 10]
```diff
@@ -3165,18 +3165,26 @@ int verity_dissect_and_mount(
...
         int r;
 
         assert(src);
         assert(dest);
 
+        if (options)
+                relax_extension_release_check =
+                        string_contains_word(mount_options_from_designator(options, PARTITION_ROOT), ",", "x-systemd.extension-release.nonstrict") ||
+                        string_contains_word(mount_options_from_designator(options, PARTITION_USR), ",", "x-systemd.extension-release.nonstrict") ||
+                        string_contains_word(options->options, ",", "x-systemd.extension-release.nonstrict");
```
Comment (poettering) [inline src/shared/dissect-image.c]: `x-systemd.relax-extension-release-check`?

Comment (poettering) [reply src/shared/dissect-image.c]: I'd prefer if this would take a boolean btw. (at least optionally), i.e `x-systemd.relax-extension-release-check=1` or so.

Also, this expression is super complex. Please add helper for it `mount_options_relax_extension_release_check()` or so that hides it

File: src/shared/mount-util.c [hunk 11]
```diff
@@ -765,7 +765,9 @@ int mount_option_mangle(
                 }
 
                 /* If 'word' is not a mount flag, then store it in '*ret_remaining_options'. */
-                if (!ent->name && !strextend_with_separator(&ret, ",", word))
+                if (!ent->name &&
+                    !streq(word, "x-systemd.extension-release.nonstrict") &&
```
Comment (poettering) [inline src/shared/mount-util.c]: why drop this here? it shouldn't hurt if we pass this along? and might be good for debugging to let it propagate this on. libmount should eat it up once passing this to the kernel, no?

File: src/portable/portable.c [hunk 12]
```diff
@@ -1004,7 +1010,16 @@ static int install_chroot_dropin(
 
                 if (m->image_path && !path_equal(m->image_path, image_path))
                         ORDERED_HASHMAP_FOREACH(ext, extension_images)
-                                if (!strextend(&text, extension_setting_from_image(ext->type), ext->path, "\n"))
+                                if (!strextend(&text,
+                                               extension_setting_from_image(ext->type),
+                                               ext->path,
+                                               /* With --force tell PID1 to avoid enforcing that the image <name> and
+                                                * extension-release.<name> have to match. */
+                                               !IN_SET(type, IMAGE_DIRECTORY, IMAGE_SUBVOLUME) &&
```
Comment (poettering) [inline src/portable/portable.c:1018]: why restrict this by image type? i see nothing bad if we allow this for other image types too?

File: src/shared/discover-image.c [hunk 13]
```diff
@@ -1174,7 +1174,7 @@ int image_read_metadata(Image *i) {
                 if (r < 0)
                         log_debug_errno(r, "Failed to read os-release in image, ignoring: %m");
 
-                r = load_extension_release_pairs(i->path, i->name, &extension_release);
+                r = load_extension_release_pairs(i->path, i->name, false, &extension_release);
```
Comment (poettering) [inline src/shared/discover-image.c]: `/* relax_…= */`

File: src/shared/dissect-image.c [hunk 14]
```diff
@@ -2714,7 +2714,7 @@ int dissected_image_acquire_metadata(DissectedImage *m, DissectImageFlags extra_
                                  * we allow a fallback that matches on the first extension-release
                                  * file found in the directory, if one named after the image cannot
                                  * be found first. */
-                                r = open_extension_release(t, m->image_name, NULL, &fd);
+                                r = open_extension_release(t, m->image_name, false, NULL, &fd);
```
Comment (poettering) [inline src/shared/dissect-image.c]: ditto

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/shared/mount-util.c (hunk 11 above)
Comment (bluca) [reply src/shared/mount-util.c]: nope, the mount syscall fails if passed through

File: src/basic/os-util.c (hunk 6 above)
Comment (bluca) [reply src/basic/os-util.c]: that's not merged yet

File: src/shared/dissect-image.h (hunk 9 above)
Comment (bluca) [reply src/shared/dissect-image.h]: shortened to SYSEXT to avoid rebreaking the table

File: src/shared/dissect-image.c (hunk 10 above)
Comment (bluca) [reply src/shared/dissect-image.c]: why a boolean? other x-systemd mount options don't do that no?

File: src/portable/portable.c (hunk 12 above)
Comment (bluca) [reply src/portable/portable.c:1018]: it's a mount option? they are not supported for the other types, there's nowhere to pass them

File: src/shared/dissect-image.c (hunk 10 above)
Comment (poettering) [reply src/shared/dissect-image.c]: you are right. ignore me then.

Comment (poettering) [reply src/shared/dissect-image.c]: (still: please split this out into a helper func)

File: src/shared/mount-util.c (hunk 11 above)
Comment (poettering) [reply src/shared/mount-util.c]: oh, hmm, this is not going through libmount here. hmm. ok.

but then maybe just ignore everything that starts with `x-` and `X-` maybe. To match at least one tiny subset of what libmount wuld do.

File: src/shared/dissect-image.c (hunk 10 above)
Comment (bluca) [reply src/shared/dissect-image.c]: yes that's already done

File: src/shared/mount-util.c (hunk 11 above)
Comment (bluca) [reply src/shared/mount-util.c]: I'm a bit wary of introducing regressions, so I'll ignore x-systemd* rather than x*, we shouldn't be using anything else in our codebase anyway

Comment (poettering) [reply src/shared/mount-util.c]: it's official libmount API, and documented. It should be safe to filter `x-*.`/`X-*.`, because that's documented clearly as userspace mount options and is filtered by libmount too.

Comment (bluca) [reply src/shared/mount-util.c]: ok, updated

File: man/org.freedesktop.portable1.xml [hunk 15]
```diff
@@ -309,7 +309,8 @@ node /org/freedesktop/portable1 {
       <function>ReattachImageWithExtensions()</function> methods take in options as flags instead of
       booleans to allow for extendability. <varname>SD_SYSTEMD_PORTABLE_FORCE</varname> will cause
       safety checks that ensure the units are not running while the new image is attached or detached
-      to be skipped. They are defined as follows:</para>
+      to be skipped, and do not insist that the extension-release.NAME file in the extension image has to
```
Comment (poettering) [inline man/org.freedesktop.portable1.xml]: as mentioned earlier, could use some `<filename>` and `<replaceable>` tag love

File: man/portablectl.xml [hunk 16]
```diff
@@ -378,7 +378,8 @@
         <term><option>--force</option></term>
 
         <listitem><para>Skip safety checks and attach or detach images (with extensions) without first ensuring
-        that the units are not running.</para></listitem>
+        that the units are not running, and do not insist that the <filename>extension-release.NAME</filename>
```
Comment (poettering) [inline man/portablectl.xml]: ditto

**Comment by poettering** [comment]
lgtm. would still love for this to be merged after https://github.com/systemd/systemd/pull/24947, so that that the new macro can be used.

File: man/portablectl.xml (hunk 16 above)
Comment (bluca) [reply man/portablectl.xml]: I've added `<filename>` here?

File: man/org.freedesktop.portable1.xml (hunk 15 above)
Comment (bluca) [reply man/org.freedesktop.portable1.xml]: I'll add `filename`, but what should `replaceable` be used for?

Comment (poettering) [reply man/org.freedesktop.portable1.xml]: NAME

**Comment by poettering** [comment]
https://github.com/systemd/systemd/pull/24947 is merged now. please rebase and make use of the new macro then.

**Comment by bluca** [comment]
> #24947 is merged now. please rebase and make use of the new macro then.

done

**Comment by bluca** [comment]
@keszybz given you asked for this change, are you happy with how it looks?

File: src/basic/os-util.c [hunk 17]
```diff
@@ -58,7 +58,47 @@ int path_is_extension_tree(const char *path, const char *extension) {
...
+
+        /* Explicitly set to request strict matching? Skip it. */
+        r = parse_boolean(extension_release_xattr);
+        if (r < 0)
+                log_debug_errno(r,
+                                "%s/%s: Failed to parse 'user.extension-release.strict' extended attribute from file, ignoring: %m",
+                                extension_release_dir_path, filename);
+        else if (r > 0)
+                log_debug("%s/%s: 'user.extension-release.strict' attribute is true, ignoring file.",
+                        extension_release_dir_path, filename);
```
Comment (yuwata) [inline src/basic/os-util.c]: Indentation?

File: src/basic/os-util.c [hunk 18]
```diff
@@ -58,7 +58,47 @@ int path_is_extension_tree(const char *path, const char *extension) {
...
+        if (r != 0)
+                return r;
+
+        log_debug("%s/%s: 'user.extension-release.strict' attribute is false%s",
+                extension_release_dir_path, filename,
+                special_glyph(SPECIAL_GLYPH_ELLIPSIS));
```
Comment (yuwata) [inline src/basic/os-util.c]: Also here.

File: src/basic/os-util.c [hunk 19]
```diff
@@ -58,7 +58,47 @@ int path_is_extension_tree(const char *path, const char *extension) {
...
+                else
+                        log_debug("%s/%s does not have user.extension-release.strict xattr, ignoring.", extension_release_dir_path, filename);
+
+                return r;
+        }
+
+        /* Explicitly set to request strict matching? Skip it. */
+        r = parse_boolean(extension_release_xattr);
+        if (r < 0)
+                log_debug_errno(r,
```
Comment (yuwata) [inline src/basic/os-util.c]: Not a matter, but you can use `return log_debug_errno()` here.

File: src/basic/os-util.c [hunk 20]
```diff
@@ -58,7 +58,47 @@ int path_is_extension_tree(const char *path, const char *extension) {
...
+        assert(extension_release_fd >= 0);
+        assert(extension_release_dir_path);
+        assert(filename);
+
+        /* No xattr or cannot parse it? Then skip this. */
+        _cleanup_free_ char *extension_release_xattr = NULL;
+        r = fgetxattr_malloc(extension_release_fd, "user.extension-release.strict", &extension_release_xattr);
+        if (r < 0) {
+                if (!ERRNO_IS_XATTR_ABSENT(r))
+                        log_debug_errno(r,
```
Comment (yuwata) [inline src/basic/os-util.c]: `return log_debug_errno()`

File: src/basic/os-util.c [hunk 21]
```diff
@@ -58,7 +58,47 @@ int path_is_extension_tree(const char *path, const char *extension) {
...
+                                        "%s/%s: Failed to read 'user.extension-release.strict' extended attribute from file, ignoring: %m",
+                                        extension_release_dir_path, filename);
+                else
+                        log_debug("%s/%s does not have user.extension-release.strict xattr, ignoring.", extension_release_dir_path, filename);
+
+                return r;
```
Comment (yuwata) [inline src/basic/os-util.c]: Then, please return `false` here (or `true`??).

File: src/basic/os-util.c [hunk 22]
```diff
@@ -58,7 +58,47 @@ int path_is_extension_tree(const char *path, const char *extension) {
...
+        }
+
+        /* Explicitly set to request strict matching? Skip it. */
+        r = parse_boolean(extension_release_xattr);
```
Comment (yuwata) [inline src/basic/os-util.c:81]: Maybe, it is better to introduce `fgetxattr_boolean()`.

File: src/basic/os-util.c [hunk 23]
```diff
@@ -259,7 +279,7 @@ int _parse_os_release(const char *root, ...) {
         int r;
 
         va_start(ap, root);
-        r = parse_release_internal(root, NULL, ap);
+        r = parse_release_internal(root, false, NULL, ap);
```
Comment (yuwata) [inline src/basic/os-util.c]: Please add `/* relax_... = */`

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/basic/os-util.c (hunk 22 above)
Comment (bluca) [reply src/basic/os-util.c:81]: that seems overkill to me, it's used only here

File: src/basic/os-util.c (hunk 21 above)
Comment (yuwata) [reply src/basic/os-util.c]: Not addressed yet?

File: src/basic/os-util.c (hunk 22 above)
Comment (yuwata) [reply src/basic/os-util.c:81]: OK, let's do that in a later PR if it helps.

File: src/basic/os-util.c (hunk 21 above)
Comment (bluca) [reply src/basic/os-util.c]: this should not return true or false, it should return the enoent error as-is

Comment (yuwata) [reply src/basic/os-util.c]: So, the failure here is also critical? Then, I'd like to drop ", ignoring" from the log message.

Comment (bluca) [reply src/basic/os-util.c]: "ignoring" here refers to the extension-release file - it will be ignored because it does not have the attribute

Comment (yuwata) [reply src/basic/os-util.c]: OK. Then, how about to explicitly say `"ignoring the extension-release file"`?

Comment (bluca) [reply src/basic/os-util.c]: it already adds the full path and filename as the first thing in the debug message

Comment (yuwata) [reply src/basic/os-util.c]: Yeah, I know. But still I'd like that.

But, not a strong opinion. I just confused when I read the code.

Comment (yuwata) [reply src/basic/os-util.c]: BTW, you can also use `return log_debug_errno()` here if you really want to propagate the error.

Comment (yuwata) [reply src/basic/os-util.c]: Hmm, still I am confused. In the below, the function returns with `true`, and `"ignoring file"`. If the failure here results the same, why not to return and log in the same way?

Comment (bluca) [reply src/basic/os-util.c]: because true means it's set and true, enoent means it's not set at all

**Comment by yuwata** [comment]
Still I am not convinced about or confused by the error handling. But hopefully that's not a big matter.

Setting the green label again, and @keszybz please take a final review.

File: man/org.freedesktop.portable1.xml [hunk 24]
```diff
@@ -309,7 +309,9 @@ node /org/freedesktop/portable1 {
       <function>ReattachImageWithExtensions()</function> methods take in options as flags instead of
       booleans to allow for extendability. <varname>SD_SYSTEMD_PORTABLE_FORCE</varname> will cause
       safety checks that ensure the units are not running while the new image is attached or detached
-      to be skipped. They are defined as follows:</para>
+      to be skipped, and do not insist that the
+      <filename>extension-release.<replaceable>NAME</replaceable></filename> file in the extension image
+      has to match the image filename. They are defined as follows:</para>
```
Comment (keszybz) [inline man/org.freedesktop.portable1.xml]: Grammar is very inconsistent between the two parts of the sentence. Maybe something like this instead:
`In addition, the check that the <filename>extension-release.<replaceable>NAME</replaceable></filename> file in the extension image matches the image name will be skipped.`

File: man/systemd.exec.xml [hunk 25]
```diff
@@ -441,7 +441,9 @@
         <para>Each image must carry a <filename>/usr/lib/extension-release.d/extension-release.IMAGE</filename>
         file, with the appropriate metadata which matches <varname>RootImage=</varname>/<varname>RootDirectory=</varname>
         or the host. See:
-        <citerefentry><refentrytitle>os-release</refentrytitle><manvolnum>5</manvolnum></citerefentry>.</para>
+        <citerefentry><refentrytitle>os-release</refentrytitle><manvolnum>5</manvolnum></citerefentry>.
+        To disable the safety check that enforces that the extension-release filename has to match the image
+        filename, the <varname>x-systemd.relax-extension-release-check</varname> mount option may be appended.</para>
```
Comment (keszybz) [inline man/systemd.exec.xml]: "To disable the safety check that enforces that the extension-release filename has to match the image filename" → "To disable the safety check that the extension-release file name matches the image file name"

File: src/basic/os-util.c [hunk 26]
```diff
@@ -58,7 +58,47 @@ int path_is_extension_tree(const char *path, const char *extension) {
         return 1;
 }
 
-int open_extension_release(const char *root, const char *extension, char **ret_path, int *ret_fd) {
+static int check_extension_release_xattr(int extension_release_fd, char *extension_release_dir_path, char *filename) {
```
Comment (keszybz) [inline src/basic/os-util.c]: Could this be renamed so that it's clear what the return value means,
e.g. `extension_release_strict_xattr_value()`? The calling convention for functions that have "check" in the name and return a boolean can be easily misunderstood.

File: src/shared/mount-util.c [hunk 27]
```diff
@@ -765,7 +765,9 @@ int mount_option_mangle(
                 }
 
                 /* If 'word' is not a mount flag, then store it in '*ret_remaining_options'. */
-                if (!ent->name && !strextend_with_separator(&ret, ",", word))
+                if (!ent->name &&
+                    !startswith_no_case(word, "x-") &&
+                    !strextend_with_separator(&ret, ",", word))
```
Comment (keszybz) [inline src/shared/mount-util.c:770]: This should be a separate commit with justification…

File: src/basic/os-util.c (hunk 21 above)
Comment (keszybz) [reply src/basic/os-util.c]: I'm with Yu on this… Saying "ignoring" and then returning error in some cases, and positive in another feels wrong. Overall, I think it'd be better if this function didn't say "ignoring" at all, since that's up to the caller.

File: src/basic/os-util.c [hunk 28]
```diff
@@ -123,34 +163,14 @@ int open_extension_release(const char *root, const char *extension, char **ret_p
...
-                                        log_debug_errno(k,
-                                                        "%s/%s: Failed to read 'user.extension-release.strict' extended attribute from file: %m",
-                                                        extension_release_dir_path, de->d_name);
-                                if (k < 0) {
-                                        log_debug("%s/%s does not have user.extension-release.strict xattr, ignoring.", extension_release_dir_path, de->d_name);
-                                        continue;
+                                if (!relax_extension_release_check) {
+                                        k = check_extension_release_xattr(extension_release_fd,
+                                                                          extension_release_dir_path,
+                                                                          de->d_name);
```
Comment (keszybz) [inline src/basic/os-util.c]: Shouldn't the splitting  out of the helper be a separate commit?

**Review by keszybz — COMMENTED** [review]
I'm not convinced that this should all be stuffed under `--force`. But also I don't have much experience with portable images, so my opinion is very weak.

The patch looks like it'll do what it says on the tin, so in this regard it seems fine. I only had some comments about style.

File: src/basic/os-util.c (hunk 21 above)
Comment (bluca) [reply src/basic/os-util.c]: This is already very complicated for my users, and a major source of support requests. I do not want to change the log messages, as that will make things even more confusing. This is just a static helper, if the log messages don't fit with it, I'll merge it back into the caller.

**Comment by bluca** [comment]
> I'm not convinced that this should all be stuffed under `--force`. But also I don't have much experience with portable images, so my opinion is very weak.

We could have separate D-Bus flag values, but for the command line tool to me it makes sense to use the same switch. We have various 'safety' checks, and with that flag you ask to bypass them. It seems pretty clear behaviour to me, and it does the expected thing.

**Comment by keszybz** [comment]
> > I'm not convinced that this should all be stuffed under `--force`. But also I don't have much experience with portable images, so my opinion is very weak.
> 
> We could have separate D-Bus flag values, but for the command line tool to me it makes sense to use the same switch. We have various 'safety' checks, and with that flag you ask to bypass them. It seems pretty clear behaviour to me, and it does the expected thing.

Yeah, that makes sense. A separate D-Bus flag will also make it easier to change behaviour in the future if we want to.

**Comment by bluca** [comment]
> > > I'm not convinced that this should all be stuffed under `--force`. But also I don't have much experience with portable images, so my opinion is very weak.
> > 
> > 
> > We could have separate D-Bus flag values, but for the command line tool to me it makes sense to use the same switch. We have various 'safety' checks, and with that flag you ask to bypass them. It seems pretty clear behaviour to me, and it does the expected thing.
> 
> Yeah, that makes sense. A separate D-Bus flag will also make it easier to change behaviour in the future if we want to.

Ok, renamed the old one added the new one.

File: src/basic/os-util.c (hunk 21 above)
Comment (keszybz) [reply src/basic/os-util.c]: Makes sense.

File: man/org.freedesktop.portable1.xml [hunk 29]
```diff
@@ -309,11 +309,14 @@ node /org/freedesktop/portable1 {
       <function>ReattachImageWithExtensions()</function> methods take in options as flags instead of
       booleans to allow for extendability. <varname>SD_SYSTEMD_PORTABLE_FORCE_ATTACH</varname> will cause
       safety checks that ensure the units are not running while the new image is attached or detached
-      to be skipped. They are defined as follows:</para>
+      to be skipped. <varname>SD_SYSTEMD_PORTABLE_FORCE_SYSEXT</varname> will cause the check that the
+      <filename>extension-release.<replaceable>NAME</replaceable></filename> file in the extension image
+      matches the image name will be skipped. They are defined as follows:</para>
```
Comment (keszybz) [inline man/org.freedesktop.portable1.xml]: to be skipped

**Review by keszybz — COMMENTED** [review]
LGTM.

File: src/basic/os-util.c [hunk 30]
```diff
@@ -58,6 +58,46 @@ int path_is_extension_tree(const char *path, const char *extension) {
...
+        r = fgetxattr_malloc(extension_release_fd, "user.extension-release.strict", &extension_release_xattr);
+        if (r < 0) {
+                if (!ERRNO_IS_XATTR_ABSENT(r))
+                        return log_debug_errno(r,
+                                               "%s/%s: Failed to read 'user.extension-release.strict' extended attribute from file, ignoring: %m",
+                                               extension_release_dir_path, filename);
+
+                log_debug("%s/%s does not have user.extension-release.strict xattr, ignoring.", extension_release_dir_path, filename);
+
+                return r;
```
Comment (poettering) [inline src/basic/os-util.c]: merge into `return log_debug_errno(r, …`?

File: src/basic/os-util.c [hunk 31]
```diff
@@ -58,6 +58,46 @@ int path_is_extension_tree(const char *path, const char *extension) {
         return 1;
 }
 
+static int extension_release_strict_xattr_value(int extension_release_fd, char *extension_release_dir_path, char *filename) {
```
Comment (poettering) [inline src/basic/os-util.c]: `char*` → `const char*`

(twice)
